### PR TITLE
bugfix: remove cnf_rewrite in push_down_filter

### LINKED ISSUE #4822: PushdownFilter rule exist bug will cause filter change wrong
**Describe the bug**

before
```sql
Projection: part.p_partkey, SUM(lineitem.l_extendedprice), AVG(lineitem.l_discount), COUNT(DISTINCT partsupp.ps_suppkey)                                                                                                                         |
  Aggregate: groupBy=[[part.p_partkey]], aggr=[[SUM(lineitem.l_extendedprice), AVG(lineitem.l_discount), COUNT(DISTINCT partsupp.ps_suppkey)]]                                                                                                   |
    Filter: part.p_partkey = lineitem.l_partkey AND part.p_brand = Utf8("Brand#12") AND part.p_partkey = partsupp.ps_partkey OR part.p_partkey = lineitem.l_partkey AND part.p_brand = Utf8("Brand#23") AND partsupp.ps_partkey = part.p_partkey |
      CrossJoin:                                                                                                                                                                                                                                 |
        CrossJoin:                                                                                                                                                                                                                               |
          TableScan: lineitem                                                                                                                                                                                                                    |
          TableScan: part                                                                                                                                                                                                                        |
        TableScan: partsupp   
```

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #4822.

# Rationale for this change

I found this error when I finish #4465

# What changes are included in this PR?

- remove cnf_rewrite (its implementation is wrong, and we already have rule `RewriteDisjunctivePredicate`)
- dedup by using hashset when merge filter 
- move rule RewriteDisjunctivePredicate into front

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sql/joins.rs [hunk 1]
```diff
@@ -1561,7 +1561,7 @@ async fn reduce_left_join_2() -> Result<()> {
         let expected = vec![
             "Explain [plan_type:Utf8, plan:Utf8]",
             "  Projection: t1.t1_id, t1.t1_name, t1.t1_int, t2.t2_id, t2.t2_name, t2.t2_int [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, t2_id:UInt32;N, t2_name:Utf8;N, t2_int:UInt32;N]",
-            "    Filter: t2.t2_int < UInt32(10) OR t1.t1_int > UInt32(2) [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, t2_id:UInt32;N, t2_name:Utf8;N, t2_int:UInt32;N]",
+            "    Filter: t2.t2_int < UInt32(10) OR t1.t1_int > UInt32(2) AND t2.t2_name != Utf8(\"w\") [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, t2_id:UInt32;N, t2_name:Utf8;N, t2_int:UInt32;N]",
```
Comment (jackwener) [inline datafusion/core/tests/sql/joins.rs:1564]: we can find `AND t2.t2_name != Utf8(\"w\")` is lost due to original `cnf_rewrite()`

**Comment by jackwener** [comment]
I prepare to remove the `cnf_rewrite` in an independent PR. Mix in a PR will make harder to revert and read.

**Comment by jackwener** [comment]
PTAL @alamb @Ted-Jiang

File: datafusion/optimizer/src/push_down_filter.rs [hunk 2]
```diff
@@ -689,19 +700,15 @@ impl OptimizerRule for PushDownFilter {
...
-
+                let predicates = utils::split_conjunction_owned(filter.predicate.clone());
                 push_down_all_join(predicates, &filter.input, left, right, vec![])?
             }
             LogicalPlan::TableScan(scan) => {
                 let mut new_scan_filters = scan.filters.clone();
                 let mut new_predicate = vec![];
 
-                let filter_predicates = utils::split_conjunction_owned(
-                    utils::cnf_rewrite(filter.predicate.clone()),
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:703]: I believe this is the only place that calls `cnf_rewrite` 

  https://github.com/search?q=repo%3Aapache%2Farrow-datafusion%20cnf_rewrite&type=code

so if we are going to stop calling it, I think we should also remove `cnf_rewrite` from the code

**Review by alamb — COMMENTED** [review]
Thanks @jackwener  -- I think this rewrite was added by @Ted-Jiang  in https://github.com/apache/arrow-datafusion/pull/3903 so perhaps he can weigh in on this proposal

Given it fixes a bug I am inclined to merge it in but will wait for @Ted-Jiang for a few days before doing so


I found https://github.com/apache/arrow-datafusion/pull/4825/files?w=1 shows the diffs much easier

**Review by alamb — APPROVED** [review]
I'll plan to merge this over the weekend unless we hear something from @Ted-Jiang

**Comment by alamb** [comment]
Thanks again @jackwener

**Pull request merged**